File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/UseItemPresenter.vb
Imports Sakuno.ING.Game.Models.Knowledge

Namespace Controls
    <TemplatePart(Name:=UseItemPresenter.Name_PART_Difference, Type:=GetType(TextBlock))>
    Public NotInheritable Class UseItemPresenter
        Inherits Control

        Friend Const Name_PART_Difference = "PART_Difference"

        Public Shared ReadOnly IdProperty As DependencyProperty =
            DependencyProperty.Register(NameOf(Id), GetType(KnownUseItem), GetType(UseItemPresenter), New PropertyMetadata(New KnownUseItem))
        Public Property Id As KnownUseItem
            Get
                Return CType(GetValue(IdProperty), KnownUseItem)
            End Get
            Set(value As KnownUseItem)
                SetValue(IdProperty, value)
            End Set
        End Property

        Public Shared ReadOnly AmountProperty As DependencyProperty =
            DependencyProperty.Register(NameOf(Amount), GetType(Integer), GetType(UseItemPresenter),
                                        New PropertyMetadata(0, Sub(d, e) CType(d, UseItemPresenter).Update(CInt(e.OldValue), CInt(e.NewValue))))
        Public Property Amount As Integer
            Get
                Return CType(GetValue(AmountProperty), Integer)
            End Get
            Set(value As Integer)
                SetValue(AmountProperty, value)
            End Set
        End Property

        Private _updateNestingCount As Integer

        Private _difference As Integer
        Private _differenceText As TextBlock

        Public Sub New()
            DefaultStyleKey = GetType(UseItemPresenter)
        End Sub

        Protected Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            _differenceText = TryCast(GetTemplateChild(Name_PART_Difference), TextBlock)
        End Sub

        Private Sub Update(oldValue As Integer, newValue As Integer)
            _difference += newValue - oldValue
            _differenceText.Text = _difference.ToString("+0;-0;0")

            UpdateVisualState()
        End Sub

        Private Async Sub UpdateVisualState()
            Const ThrottleTime = 3000

            VisualStateManager.GoToState(Me, "ValueChanged", True)

            _updateNestingCount = _updateNestingCount + 1
            Await Task.Delay(ThrottleTime)
            _updateNestingCount = _updateNestingCount - 1

            If _updateNestingCount > 0 Then
                Return
            End If

            VisualStateManager.GoToState(Me, "Normal", True)

            _difference = 0
            _differenceText.Text = String.Empty
        End Sub

    End Class
End Namespace

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/UseItemPresenter.cs
using System;
using System.Threading.Tasks;
using Sakuno.ING.Game.Models.Knowledge;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace Sakuno.ING.Views.UWP.Controls
{
    [TemplatePart(Name = nameof(PART_Difference), Type = typeof(TextBlock))]
    public class UseItemPresenter : Control
    {
        public static readonly DependencyProperty IdProperty
            = DependencyProperty.Register(nameof(Id), typeof(KnownUseItem), typeof(UseItemPresenter), new PropertyMetadata(default(KnownUseItem)));

        public KnownUseItem Id
        {
            get => (KnownUseItem)GetValue(IdProperty);
            set => SetValue(IdProperty, value);
        }

        public static readonly DependencyProperty AmountProperty =
            DependencyProperty.Register(nameof(Amount), typeof(int), typeof(UseItemPresenter),
                new PropertyMetadata(0, (d, e) => ((UseItemPresenter)d).Update((int)e.OldValue, (int)e.NewValue)));

        public int Amount
        {
            get => (int)GetValue(AmountProperty);
            set => SetValue(AmountProperty, value);
        }

        public UseItemPresenter()
        {
            DefaultStyleKey = typeof(UseItemPresenter);
        }

        private TextBlock PART_Difference;
        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            PART_Difference = GetTemplateChild(nameof(PART_Difference)) as TextBlock;
        }

        private int _difference;
        private void Update(int oldValue, int newValue)
        {
            _difference += newValue - oldValue;
            PART_Difference.Text = _difference.ToString("+0;-0;0");

            UpdateVisualState();
        }

        private int _updateNestingCount;
        private async void UpdateVisualState()
        {
            const int ThrottleTime = 3000;

            VisualStateManager.GoToState(this, "ValueChanged", true);

            _updateNestingCount++;
            await Task.Delay(ThrottleTime);
            _updateNestingCount--;

            if (_updateNestingCount > 0)
                return;

            VisualStateManager.GoToState(this, "Normal", true);

            _difference = 0;
            PART_Difference.Text = string.Empty;
        }
    }
}
